### DISCUSSION TITLE
YoloV8 model still not running in a GPU

### DISCUSSION THREAD

**Discussion by neliel123** [ultralytics/ultralytics#14251] [Q&A]
Hello,

My problem is common and I have tried all troubleshooting that I could find here but I am still not able to run my YoloV8 model to run in a GPU

I am still getting "Using CPU. Note: This module is much faster with a GPU."

I have checked my CUDA installation on my Windows machine and below is the output

```
>>> import torch
>>> print(torch.cuda.device_count())
1
>>> print(torch.cuda.get_device_name()) 
NVIDIA GeForce GTX 1650
```

And this is my code.

```
import cv2
import easyocr
from ultralytics import YOLO
import asyncio
import aiohttp
from datetime import datetime, timedelta
from utils import write_csv, read_license_plate
from helper import create_video_writer
import torch
import time

async def insert_plate_data(session, data):
    pass

async def process_frame(frame, frame_nmr, license_plate_model, reader, ignored_texts, last_seen):
    frame_results = []
    detections = license_plate_model(frame)[0]
    
    for detection in detections.boxes.data.tolist():
        x1, y1, x2, y2, confidence, class_id = detection
        license_plate_crop = frame[int(y1):int(y2), int(x1):int(x2)]
        license_plate = read_license_plate(license_plate_crop)

        if license_plate:
            now = datetime.now()
            if license_plate not in last_seen or now - last_seen[license_plate] > timedelta(seconds=10):
                frame_results.append({
                    'plate_number': license_plate
                })
                last_seen[license_plate] = now

    return frame_results

async def main():
    # Set up camera connection details
    ip_address = '192.168.100.11' # Replace with the IP address of your camera
    port = '554'          # Replace with the port number for your camera
    username = 'camera_admin'    # Replace with the username for your camera
    password = 'WifiCamera123!' # Replace with the password for your camera

    # Construct the RTSP stream URLs using variables
    url_640x480 = f"rtsp://{username}:{password}@{ip_address}:{port}/stream2"
    high_quality_stream = f"rtsp://{username}:{password}@{ip_address}:{port}/stream1"

    # Set up RTSP stream URL
    rtsp_url = high_quality_stream
    video_cap = cv2.VideoCapture(rtsp_url)

    device = 'cuda' if torch.cuda.is_available() else 'cpu'
    print(f"Using device {device}")

    # Initialize YOLO model
    license_plate_model = YOLO('./models/last.pt').to(device)
    reader = easyocr.Reader(['en'])

    frame_nmr = -1
    ignored_texts = ['registered', 'region', 'temporary']
    last_seen = {}  
    prev_time = time.time()
    
    async with aiohttp.ClientSession() as session:
        while True:
            frame_nmr += 1
            ret, frame = video_cap.read()
            
            if not ret:
                break

            # Calculate FPS
            current_time = time.time()
            fps = 1 / (current_time - prev_time)
            prev_time = current_time
            
            frame_results = await process_frame(frame, frame_nmr, license_plate_model, reader, ignored_texts, last_seen)
            
            if frame_results:
                for result in frame_results:
                    await insert_plate_data(session, result)
            
            # Position the FPS label on the right side of the screen
            frame_height, frame_width = frame.shape[:2]
            fps_text = f"FPS: {fps:.2f}"
            text_size, _ = cv2.getTextSize(fps_text, cv2.FONT_HERSHEY_SIMPLEX, 2, 8)
            text_x = frame_width - text_size[0] - 50
            text_y = 50
            cv2.putText(frame, fps_text, (text_x, text_y), cv2.FONT_HERSHEY_SIMPLEX, 2, (0, 0, 255), 8)
            cv2.imshow("Frame", frame)
            # writer.write(frame)
            
            if cv2.waitKey(1) == ord("q"):
                break

    video_cap.release()
    cv2.destroyAllWindows()

if __name__ == "__main__":
    asyncio.run(main())

```

When I execute my code and run nvidia-smi on a separate terminal, I am getting this
```
Mon Jul  8 09:14:03 2024
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 522.06       Driver Version: 522.06       CUDA Version: 11.8     |
|-------------------------------+----------------------+----------------------+
| GPU  Name            TCC/WDDM | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  NVIDIA GeForce ... WDDM  | 00000000:01:00.0 Off |                  N/A |
| N/A   53C    P8     1W /  N/A |   1274MiB /  4096MiB |      7%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|    0   N/A  N/A     11824      C   ...ython\Python39\python.exe    N/A      |
+-----------------------------------------------------------------------------+
```

Not sure what is missing in my setup for me to be able to run my YoloV8 model in a GPU.

Any idea?

**Comment by neliel123** [comment]
The following are the libraries that I have installed on my project using my Windows 11 laptop
```
opencv-contrib-python==4.10.0.84
opencv-python==4.10.0.84
torch==2.3.1+cu118
torchaudio==2.3.1+cu118
torchvision==0.18.1+cu118
```
tried this code
```
    device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
    print(f"Using device {device}")

    # Initialize YOLO model
    license_plate_model = YOLO('./models/last.pt').to(device)
```

Still prints
```
Using CPU. Note: This module is much faster with a GPU
Using device cuda
```

Tried everything [here](https://github.com/ultralytics/ultralytics/issues/3084)

Any hints or help please?

**Reply by pderrenger** [reply]
Hello,

Thank you for providing detailed information about your setup and the steps you've already taken to resolve the issue. It looks like you've done a thorough job troubleshooting so far. Let's try a few more steps to ensure your YOLOv8 model runs on the GPU.

### Verify CUDA and PyTorch Compatibility
First, let's ensure that your CUDA installation is compatible with your PyTorch version. You can verify this by running the following commands in your Python environment:

```python
import torch
print(torch.version.cuda)  # Should match your CUDA version
print(torch.backends.cudnn.version())  # Should return a valid cuDNN version
print(torch.cuda.is_available())  # Should return True
```

### Check YOLOv8 GPU Utilization
To ensure that YOLOv8 is utilizing the GPU, you can modify your code slightly to print the device being used by the model:

```python
import torch
from ultralytics import YOLO

# Check if CUDA is available and set the device
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
print(f"Using device {device}")

# Initialize YOLO model and move it to the device
license_plate_model = YOLO('./models/last.pt').to(device)
print(f"Model is on device: {next(license_plate_model.model.parameters()).device}")
```

### Ensure Correct Environment Activation
Make sure you are running your script in the correct virtual environment where all the necessary packages are installed. Sometimes, the issue can be as simple as running the script in a different environment.

### Update Ultralytics and Dependencies
Ensure you are using the latest versions of the Ultralytics package and its dependencies. You can update them using pip:

```bash
pip install --upgrade ultralytics torch torchvision torchaudio
```

### Additional Debugging Steps
If the issue persists, consider running a minimal reproducible example to isolate the problem. You can find guidance on creating one here: [Minimum Reproducible Example](https://docs.ultralytics.com/help/minimum_reproducible_example).

### Example Code for Debugging
Here is a minimal example to test if the model runs on the GPU:

```python
import torch
from ultralytics import YOLO

# Check if CUDA is available and set the device
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
print(f"Using device {device}")

# Initialize YOLO model and move it to the device
model = YOLO('yolov8n.pt').to(device)
print(f"Model is on device: {next(model.model.parameters()).device}")

# Dummy input to test inference
dummy_input = torch.randn(1, 3, 640, 640).to(device)
output = model(dummy_input)
print("Inference successful")
```

If the above code runs successfully and prints "Inference successful" with the model on the GPU, then the issue might be specific to your original script.

Feel free to reach out if you have any more questions or need further assistance. 😊

**Comment by neliel123** [comment]
Anyone have other advise to make my project run faster on a GPU?
Running it on a CPU is really slow.

**Reply by pderrenger** [reply]
Hello,

Thank you for sharing the details of your setup and code. It looks like you've done a thorough job of checking your CUDA installation and ensuring your environment is set up correctly. Here are a few additional steps to help ensure your YOLOv8 model runs on the GPU:

1. **Ensure CUDA and PyTorch Compatibility**:
   Make sure that your CUDA version is compatible with the version of PyTorch you have installed. You can check the compatibility matrix on the [PyTorch website](https://pytorch.org/get-started/previous-versions/).

2. **Verify GPU Usage in Code**:
   Ensure that the model and tensors are explicitly moved to the GPU. In your code, you are using `license_plate_model.to(device)`, which is correct. However, also ensure that any tensors used in the model inference are on the GPU. For example:
   ```python
   detections = license_plate_model(frame.to(device))[0]
   ```

3. **Check for Device Assignment**:
   Ensure that the device assignment is correctly set throughout your code. Sometimes, intermediate tensors might default to the CPU if not explicitly moved to the GPU.

4. **Update Packages**:
   Ensure you are using the latest versions of the Ultralytics YOLO library and PyTorch. Updates often include performance improvements and bug fixes.

5. **Monitor GPU Utilization**:
   Use `nvidia-smi` to monitor GPU utilization while your script is running. This can help identify if the GPU is being underutilized or if there are any bottlenecks.

6. **Simplify the Code for Debugging**:
   Try running a simplified version of your code to isolate the issue. For example, run a basic inference loop without the async components to see if the GPU is utilized:
   ```python
   import torch
   from ultralytics import YOLO

   device = 'cuda' if torch.cuda.is_available() else 'cpu'
   print(f"Using device {device}")

   model = YOLO('./models/last.pt').to(device)
   frame = torch.randn(1, 3, 640, 640).to(device)  # Example frame

   results = model(frame)
   print(results)
   ```

If the issue persists, please provide a minimum reproducible example as outlined [here](https://docs.ultralytics.com/help/minimum_reproducible_example). This will help us diagnose the problem more effectively.

I hope these suggestions help! Let us know how it goes or if you have any further questions. 😊
